lib/elixir/lib/map_set.ex

```diff
@@ -63,6 +63,47 @@ defmodule MapSet do
   # with Elixir 1.14 and earlier and executed on Elixir 1.15+.
   defstruct map: :sets.new(version: 2)
 
+  @doc """
+  Checks if `map_set` contains `value`. Allowed in guards.
+
+  Otherwise functions like `member?/2`.
+
+  ## Examples
+
+      iex> MapSet.is_member(MapSet.new([1, 2, 3]), 2)
+      true
+
+      iex> MapSet.is_member(MapSet.new([1, 2, 3]), 4)
+      false
+
+      iex> case MapSet.new([1, 2, 3]) do
+      ...>   map_set when MapSet.is_member(map_set, 3) -> true
+      ...>   _ -> false
+      ...> end
+      true
+
+  """
+  @doc guard: true, since: "1.17.0"
+  defmacro is_member(map_set, value) do
+    case Macro.Env.in_guard?(__CALLER__) do
+      true ->
+        quote do
+          (unquote(map_set).__struct__ == unquote(__MODULE__) or :fail) and
```

Comment:
I didn't follow why we didn't use `is_struct/1`, is there any difference from dialyzer's perspective?
```suggestion
          (is_struct(unquote(map_set), unquote(__MODULE__)) or :fail) and
```

Reply:
No difference, both implementation emit the same opaque-access dialyzer warnings and nothing more.

Reply:
`is_struct` is expanded in-place to further guards, `.__struct__` access appears to get expanded later to (I believe) the exact same guards. Compare:

```elixir
Macro.expand(quote(do: is_struct(map_set)), %Macro.Env{__ENV__ | context: :guard})
```

vs

```elixir
Macro.expand(quote(do: map_set.__struct__ == MapSet), %Macro.Env{__ENV__ | context: :guard})
```

Reply:
Ultimately I think `is_struct/2` is the best way to go about this, keeping things in sync with any changes made in that standard guard.